Linux large pages - after rebase

uttampawar opened the pull request:
Support under Linux to map the text segment of node into 2M pages.
As requested in PR #21064, I've re-base the huge page support code with master branch.

Make test results show no failures.
$ make test
[03:13|% 100|+  2363|-     0]: Done

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

uttampawar commented:
@gireeshpunathil I've rebased the code. Please let me know if I'm missing anything. Thanks.

gireeshpunathil commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16144/

gireeshpunathil commented:
@uttampawar - thanks. Given that this is opened in lieu of #21064, 
 - can we have that one closed?
 - can you confirm that the code is same as #21064 ?

uttampawar commented:
@gireeshpunathil The code is exactly the same. And yes #21064 can be closed.

uttampawar commented:
Last change (addition of new line) is the only difference between original code and the one I submitted.

gireeshpunathil commented:
@addaleax @Fishrock123 @gabrielschulhof  - this is a clone of #21064 (resubmitted here due to author unavailable I guess). Given that you had reviewed that, will you please have a look?

Also the feature under this PR is exercised only if `configured` with `--use-largepages` (that too only in Linux) so running CI does not cover that, instead only helps making sure it does not break anything. How do we go about this in general? custom CI that builds with custom params? /cc @Trott

addaleax commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,259 @@
+    return -1;
+  }
+
+  if (r.from > reinterpret_cast<void* > (&MoveTextRegionToLargePages))
```
ditto here, no space before/after the cast’s `>`

addaleax commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,259 @@
+int MapStaticCodeToLargePages() {
+  struct text_region r = FindNodeTextRegion();
+  if (r.found_text_region == false) {
+    fprintf(stderr, "Hugepages WARNING: failed to find text region \n");
```
There’s an extra space at the end of the message here

addaleax commented on src/node_large_page.h:
```diff
@@ -0,0 +1,36 @@
+
+namespace node {
+  bool IsLargePagesEnabled();
+  int MapStaticCodeToLargePages();
```
style nit: no indentation inside namespaces, blank lines/before after the namespace lines

addaleax commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,259 @@
+//    If successful copy the code there and unmap the original region.
+
+extern char __executable_start;
+extern char __etext;
```
Where are these two used?

Trott commented:
I'm not sure if this has security implications, but out of caution: @nodejs/security-wg

uttampawar commented:
@addaleax Thanks for the review. You are correct, this is re-submission since original author is on vacation. I'll address suggested changes.

uttampawar commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,259 @@
+// We want PROT_WRITE because we are writing into it.
+// We want it at the fixed address and we use MAP_FIXED.
+  tmem = mmap(start, size,
+              PROT_READ | PROT_WRITE | PROT_EXEC,
```
In a NOTES section of "man mmap" it states that, "It is architecture dependent whether PROT_READ implies PROT_EXEC or not. Portable programs should always set PROT_EXEC if they intend to execute code in the new mapping".

uttampawar commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,259 @@
+    fprintf(stderr, "Could not open /proc/self/maps\n");
+    return nregion;
+  }
+  std::getline(ifs, map_line);
```
Hmm this is strange. Normally, for every process I checked on Linux system, the first line of /proc/<pid>/maps always contain the metadata with absolute path to the executable. For example on my system, I see 
 $ ./node.orig -p 'fs.readFileSync("/proc/self/maps", "utf8")'
00400000-00c00000 r-xp 00000000 08:01 667980        /home/upawar/node-large-pages/node.orig

uttampawar replied:
@addaleax Another example with released node version,
$ ./node-v8.11.3-linux-x64/bin/node -p 'fs.readFileSync("/proc/self/maps", "utf8")'
00400000-01f4b000 r-xp 00000000 08:01 533163 /home/upawar/node-v8.11.3-linux-x64/bin/node
...

gabrielschulhof replied:
Looks like we might want to generalize this by retrieving the value of `process.execPath` and then looking for the line that is executable and ends in that value.

gabrielschulhof replied:
... which I guess we can get from [`uv_exepath()`](http://docs.libuv.org/en/v1.x/misc.html?highlight=uv_exepath#c.uv_exepath).

suresh-srinivas replied:
@addaleax I havent seen this before. According to Linux proc system documentation and manual https://www.kernel.org/doc/Documentation/filesystems/proc.txt the first entry is the executable file
```
The format is:
address           perms offset  dev   inode      pathname
08048000-08049000 r-xp 00000000 03:00 8312       /opt/test
08049000-0804a000 rw-p 00001000 03:00 8312       /opt/test
0804a000-0806b000 rw-p 00000000 00:00 0          [heap]

```
In your case it looks like some anonymous mapping exist before this. I will implement a loop to get to the first entry. with the executable. I think in almost all cases it should be the first one. 

@gabrielschulhof thanks, we can just use` readlink` on the `proc/self/exe `to get the actual path to compare.

uttampawar commented:
From the test log I see "test-trace-event-promises" test failed as shown below,
not ok 1878 parallel/test-trace-event-promises
I don't see this failure on dev machine. Is there anything can I do to address this issue? Is this failure due to my change?

addaleax commented:
> From the test log I see "test-trace-event-promises" test failed as shown below,
not ok 1878 parallel/test-trace-event-promises
I don't see this failure on my "dev" machine. Is there anything can I do to address this issue? Is this failure due to my change?

@uttampawar It’s unrelated (was due to a broken `master` branch), and it will go away after rebasing :)

uttampawar commented:
@addaleax Do I need to re-base again and resubmit?

lundibundi commented:
@uttampawar just rebase and force push to your branch (`uttampawar:intel-large-pages`), no need to resubmit, GitHub will update this PR for you.

uttampawar commented on configure:
```diff
@@ -977,6 +983,9 @@ def configure_node(o):
   else:
     o['variables']['node_use_dtrace'] = 'false'
 
+  use_large_pages = (flavor == 'linux' and options.node_use_large_pages)
```
@lundibundi I thin k it's a good idea. Failing early will help catch a build system error when Huge pages are not supported. 
@addaleax @jasnell Any other opinions?

lundibundi replied:
@uttampawar 
I think we can go ahead and implement it this way considering
* it is done this way in other places (see code below your change, enable-lto, dtrace etc)
* considered a good change at least by me, you and @gireeshpunathil.

uttampawar commented on configure:
```diff
@@ -400,6 +400,12 @@ parser.add_option('--with-etw',
+    action='store_true',
+    dest='node_use_large_pages',
+    help='build with Large Pages support (enabled only for Linux).' +
+         '(Needs Linux kernel >= 2.6.38 with Transparent Huge pages enabled)')
```
Agreed.

uttampawar commented:
@addaleax @lundibundi Finished re-base of large pages support patch.
I don't see any errors after rebasing as shown below,
$ make test
[03:24|% 100|+ 2365|-   0]: Done

$ make test-ci
...
# Clean up any leftover processes, error if found.
ps awwx | grep Release/node | grep -v grep | cat

uttampawar commented:
The build log says failure due to "no space left" error. See details below,

not ok 2294 sequential/test-fs-watch
  ---
  duration_ms: 0.111
  severity: fail
  exitcode: 1
  stack: |-
    internal/fs/watchers.js:170
        throw error;
        ^
    
    Error: ENOSPC: no space left on device, watch '/home/travis/build/nodejs/node/test/.tmp.0/watch.txt'
        at FSWatcher.start (internal/fs/watchers.js:164:26)
        at Object.watch (fs.js:1255:11)
        at Object.<anonymous> (/home/travis/build/nodejs/node/test/sequential/test-fs-watch.js:48:22)
        at Module._compile (internal/modules/cjs/loader.js:689:30)
        at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
        at Module.load (internal/modules/cjs/loader.js:599:32)
        at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
        at Function.Module._load (internal/modules/cjs/loader.js:530:3)
        at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
        at startup (internal/bootstrap/node.js:266:19)

lundibundi commented:
@uttampawar could you please rebase once again, just to be sure this PR has all of the updates from master (there were problems with the CI that has been fixed on master).

Ping @addaleax @gabrielschulhof @gireeshpunathil. Also, wdyt of https://github.com/nodejs/node/pull/22079#discussion_r209718594.

gireeshpunathil commented:
raising exception on irrelevant platforms looks like a good option to me.

lundibundi commented:
ping @uttampawar, PTAL https://github.com/nodejs/node/pull/22079#issuecomment-415985877.

uttampawar commented:
@gireeshpunathil @addaleax @gabrielschulhof Can someone start CI with --use-largepages option please?

gireeshpunathil commented:
> and CI)

but why doesn't the CI result appear here? @addaleax - I started a CI few hours back (please see the link above), it's result does not seem to be available in this page: either as a summary or details. I am just wondering whether I ran it correctly or not, will you please have a look?

uttampawar commented:
@addaleax Are you talking about https://github.com/nodejs/node/pull/22079/files/0e30f77b335cb4771658de3313337f095846bcc6#r207510675?

addaleax commented:
@uttampawar Yes, at least that + the indentation style nit

lundibundi commented:
Also, there is https://github.com/nodejs/node/pull/22079#discussion_r209718594.

addaleax commented:
@gireeshpunathil Sorry, only now seeing your comment – I don’t know if CI ran successfully. The last linked run that you posted is red, but I think when I commented the Github UI looked positive?

suresh-srinivas commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2018 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
```
@jasnell thanks. This was header provided by  Intel's Open Source Legal team for any new source code. We were following that guideline.

refack replied:
In that case, IMHO this should go into `/src/large_pages/` and that directory should have a `readme.md` and `license`

refack replied:
The Node.js license reads:
https://github.com/nodejs/node/blob/b2e4cb7b285e959438c8bdbf8edda3f69d2ffe98/LICENSE#L4
https://github.com/nodejs/node/blob/b2e4cb7b285e959438c8bdbf8edda3f69d2ffe98/LICENSE#L25-L26
https://github.com/nodejs/node/blob/b2e4cb7b285e959438c8bdbf8edda3f69d2ffe98/LICENSE#L49-L52
So we might have a conflict here.

IMHO the simplest way to resolve this is to explicitly delineate this and the .h file

lundibundi commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,257 @@
+    return -1;
+  }
+
+  if (r.from > reinterpret_cast<void*> (&MoveTextRegionToLargePages))
```
Still here https://github.com/nodejs/node/pull/22079/files#r209719772. =)

uttampawar commented:
@lundibundi I appreciate your time reviewing the changes. I've pushed two(2) more commits addressing last two concerns.

lundibundi commented on configure.py:
```diff
@@ -1013,8 +1013,15 @@ def configure_node(o):
+      o['variables']['node_use_large_pages'] = b(use_large_pages)
+    else:
+      raise Exception(
+        'Large pages is supported only on Linux Systems.')
```
s/is supported/are supported

lundibundi commented on configure.py:
```diff
@@ -1013,8 +1013,15 @@ def configure_node(o):
 
-  use_large_pages = (flavor == 'linux' and options.node_use_large_pages)
-  o['variables']['node_use_large_pages'] = b(use_large_pages)
+  if options.node_use_large_pages:
```
I think this can be simplified to
```python
if options.node_use_large_pages and flavor != 'linux':
  raise Exception(
    'Large pages are supported only on Linux Systems.')
o['variables']['node_use_large_pages'] = b(options.node_use_large_pages)
```
wdyt?

uttampawar replied:
@lundibundi There are just so many ways. This is makes it look succinct for sure. I'll make the change.

uttampawar replied:
@lundibundi Pushed new commit addressing both the suggestions. Thanks.

uttampawar commented:
@addaleax I've addressed all the concerns raised so far. Please let me know anything else you see and want me to fix. I appreciate your time reviewing our PR.

lundibundi commented on src/node_large_page.cc:
```diff
@@ -79,16 +79,18 @@ inline int64_t hugepage_align_down(int64_t addr) {
+// The format of the maps file is the following
+// address           perms offset  dev   inode       pathname
+// 00400000-00452000 r-xp 00000000 08:02 173521      /usr/bin/dbus-daemon
+// This is also handling the case w where the first line is not the binary
```
Typo `..case w where..` -> `..case where..`

lundibundi replied:
Still unresolved 

> Typo `..case w where..` -> `..case where..`

https://github.com/nodejs/node/pull/22079/files#discussion_r219667166.

lundibundi commented:
CI with `--use-largepages`: https://ci.nodejs.org/job/node-test-commit/21684/

uttampawar commented:
It looks like build failed because patch didn't merged cleanly. Shall I rebase the changes again?

lundibundi commented:
@uttampawar looks strange tbh, let's try that. I'll start the CI again after the rebase.

addaleax commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,278 @@
+
+  std::string exename;
+  {
+      char selfexe[ PATH_MAX ];
```
nit: no spaces here

suresh-srinivas replied:
resolved in the new commit

addaleax commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,278 @@
+__attribute__((__section__(".lpstub")))
+__attribute__((__aligned__(2 * 1024 * 1024)))
+__attribute__((__noinline__))
+__attribute__((__optimize__("2")))
```
Is there a reason to set a fixed optimization level here?

suresh-srinivas replied:
I added this as I wanted gcc to use O2 where no inlining of calls and no optimization of memcpy within this function happens. I am particularly concerned about the memcpy after the madvise call. If it gets replaced with something that is mapped out by the mmap call, it will crash here.

addaleax replied:
I get the reasoning for `noinline` (and maybe you want `flatten` as well?) – I’m just a bit worried that setting a fixed optimization level might be a bit brittle?

suresh-srinivas commented on src/node_large_page.cc:
```diff
@@ -98,29 +100,46 @@ static struct text_region FindNodeTextRegion() {
+    iss >> offset;
+    iss >> dev;
+    iss >> inode;
+    if (inode != 0) {
```
Yes `pathname` is available only when `inode != 0`. We can merge the other two checks.

suresh-srinivas replied:
The other two checks have now been mered

suresh-srinivas commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,259 @@
+// We already know the original page is r-xp
+// (PROT_READ, PROT_EXEC, MAP_PRIVATE)
+// We want PROT_WRITE because we are writing into it.
+// We want it at the fixed address and we use MAP_FIXED.
```
This has been addressed.

lundibundi commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,276 @@
+          nregion.to = to;
+          nregion.total_hugepages = size / hps;
+        }
+          break;
```
Unnecessary additional indentation.

lundibundi commented:
New CI with `--use-largepages`: https://ci.nodejs.org/job/node-test-commit/21771/

uttampawar commented:
@lundibundi Can you start a new CI with --use-largepages please?

lundibundi commented:
@uttampawar CI is not very stable right now (you can check last runs https://ci.nodejs.org/job/node-test-pull-request/ here on the left) so it will most likely fail for unrelated reasons (like it did in the prev run). I (or someone else) will start it later when CI becomes a bit more stable (that's why I added needs-ci label). That will most likely require a rebase, I'll ping you later for it.

lundibundi commented:
I think we can go on and try a CI =).
Please rebase on master and I will start it.
@uttampawar @suresh-srinivas

lundibundi commented:
ping @uttampawar @suresh-srinivas.

Also, @addaleax I think we can remove the red X as those changes were made.

addaleax commented:
@lundibundi Generally, when you think that my requested changes have been addressed you can dismiss them if you like :)

addaleax commented on src/node_large_page.cc:
```diff
@@ -0,0 +1,276 @@
+    if (inode != 0) {
+      std::string pathname;
+      iss >> pathname;
+      if (pathname.compare(exename) == 0 && permission.compare("r-xp") == 0) {
```
style nit: you can just use `pathname == exename && permission == "r-xp"`

uttampawar commented:
@lundibundi I'll rebase and address one more style nit as pointed by @addaleax

uttampawar commented:
@lundibundi Updated PR with rebase and style:nit. 
make test results,
[----------] Global test environment tear-down
[==========] 80 tests from 10 test cases ran. (149 ms total)
[  PASSED  ] 80 tests.

make -s jstest
[03:26|% 100|+ 2421|-   0]: Done

lundibundi commented:
New CI with largepages: https://ci.nodejs.org/job/node-test-commit/21948/

uttampawar commented:
@lundibundi Addressed the typo error as you pointed out.

lundibundi commented:
@uttampawar @suresh-srinivas 
Ok, I'm kind of confused. Somehow with `--use-largepages` flag this
* succeded on freebsd10 but failed on freebsd11 https://ci.nodejs.org/job/node-test-commit-freebsd/20895/ - configure throw of Largepages not supported
* succeded on macos1010 but failed on macos1011 https://ci.nodejs.org/job/node-test-commit-osx/21600/ - configure throw of Largepages not supported

-----------------------

Also seeing other error of 
```
/home/iojs/build/workspace/node-test-commit-linux-containered/nodes/ubuntu1604_sharedlibs_shared_x64/out/Release/obj.target/libnode.so.67: undefined reference to `__nodetext'
```
or similar with
* succeded on ppcbe-ubuntu1404 but failed on ppcle-ubuntu1404 https://ci.nodejs.org/job/node-test-commit-plinux/20625/
* succeded on every containered ubuntu but not on ubuntu1604_sharedlibs_shared_x64 https://ci.nodejs.org/job/node-test-commit-linux-containered/7491/
* succeded on 3/5 arm configs (failed on `centos7-arm64-gcc6` and `ubuntu1604-arm64`) https://ci.nodejs.org/job/node-test-commit-arm/18882/
* failed with the same on linux one https://ci.nodejs.org/job/node-test-commit-linuxone/5714/

-----------------------

Also, not sure but PTAL at this ENOMEM failure at [fedora-latest](https://ci.nodejs.org/job/node-test-commit-linux/22015/nodes=fedora-last-latest-x64/console) and [centos7](https://ci.nodejs.org/job/node-test-commit-linux/22015/nodes=centos7-64-gcc6/console)
```cmd
22:15:33 not ok 2265 sequential/test-child-process-pass-fd
22:15:33   ---
22:15:33   duration_ms: 3.321
22:15:33   severity: fail
22:15:33   exitcode: 1
22:15:33   stack: |-
22:15:33     internal/child_process.js:359
22:15:33         throw errnoException(err, 'spawn');
22:15:33         ^
22:15:33     
22:15:33     Error: spawn ENOMEM
22:15:33         at ChildProcess.spawn (internal/child_process.js:359:11)
22:15:33         at spawn (child_process.js:539:9)
22:15:33         at fork (child_process.js:109:10)
22:15:33         at forkWorker (/home/iojs/build/workspace/node-test-commit-linux/nodes/fedora-last-latest-x64/test/sequential/test-child-process-pass-fd.js:41:18)
22:15:33         at Object.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/fedora-last-latest-x64/test/sequential/test-child-process-pass-fd.js:54:5)
22:15:33         at Module._compile (internal/modules/cjs/loader.js:703:30)
22:15:33         at Object.Module._extensions..js (internal/modules/cjs/loader.js:714:10)
22:15:33         at Module.load (internal/modules/cjs/loader.js:613:32)
22:15:33         at tryModuleLoad (internal/modules/cjs/loader.js:552:12)
22:15:33         at Function.Module._load (internal/modules/cjs/loader.js:544:3)
22:15:33   ...
```

-----------------------

This on smaros17 seems irrelevant:
```cmd
22:19:42 not ok 1798 parallel/test-tls-wrap-econnreset-localaddress
22:19:42   ---
22:19:42   duration_ms: 0.420
22:19:42   severity: fail
22:19:42   exitcode: 1
22:19:42   stack: |-
22:19:42     assert.js:84
22:19:42       throw new AssertionError(obj);
22:19:42       ^
22:19:42     
22:19:42     AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
22:19:42     + actual - expected
22:19:42     
22:19:42     + 'EADDRNOTAVAIL'
22:19:42     - 'ECONNRESET'
22:19:42         at TLSSocket.tls.connect.once.common.mustCall (/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/test/parallel/test-tls-wrap-econnreset-localaddress.js:21:14)
22:19:42         at TLSSocket.<anonymous> (/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/test/common/index.js:349:15)
22:19:42         at Object.onceWrapper (events.js:273:13)
22:19:42         at TLSSocket.emit (events.js:182:13)
22:19:42         at emitErrorNT (internal/streams/destroy.js:82:8)
22:19:42         at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
22:19:42         at process._tickCallback (internal/process/next_tick.js:63:19)
22:19:42   ...
```

suresh-srinivas commented:
> @lundibundi @suresh-srinivas We will work to resolve this new conflict. Also, we think we know the reason for undefined reference of '__nodetext' error.

It looks like the way the implicit ld.script which is where `__nodetext` is defined is not used in one of the builds which is building `libnode.so`

We are creating a new rules in gypi to include it. This change is also more cleaner in node.gypi. What do you all think @lundibundi @gireeshpunathil @addaleax 
```

    [ 'OS=="linux" and target_arch=="x64" and node_use_large_pages=="true"', {
      'ldflags': [ '-Wl,-T <(PRODUCT_DIR)/../../ld.implicit.script' ]
    }], 


```

lundibundi commented:
@suresh-srinivas this looks good to me but I'm not very familiar with gyp, so maybe @addaleax may provide a better input on this one. Though a few questions:

Why was it initially set to only be built on `node_shared=="false"`?

You have put `target_arch=="x64"`, but the .gyp file only checks for `OS=="linux"` (https://github.com/nodejs/node/pull/22079/files#diff-259e627be3fafed9e9c10a7fcb26879eR627) is that correct?

suresh-srinivas commented:
> Why was it initially set to only be built on `node_shared=="false"`?
The algorithm relocated the `.text` of the `node` binary itself and not any of the dso's. So if node is embedded into another binary as a DSO (`libnode.so`) this wont apply. 
> 
> You have put `target_arch=="x64"`, but the .gyp file only checks for `OS=="linux"` (https://github.com/nodejs/node/pull/22079/files#diff-259e627be3fafed9e9c10a7fcb26879eR627) is that correct?

Good catch, we should update the `.gyp` file to include check for `x64`

lundibundi commented:
@suresh-srinivas I think it's completely okay to throw for `shared` + `use-largepages` if that's the case.
I'll start another CI once this is resolved.

Also, have you figured out why did it fail on freebsd11 but succeeded on freebsd10 etc (first point here https://github.com/nodejs/node/pull/22079#issuecomment-426386196)?

uttampawar commented:
@lundibundi @suresh-srinivas Pushed a new commit after the rebase. We have addressed following build issues, 
large pages support and non x64 platform,
large pages support and building node as a library with --shared or --enable-static option (cause for the undefined reference error for symbol "__nodetext")

Fixed node.gyp conflict. This was probably caused during one of the previous rebase (my mistake).

I don't know why the build failed on freebsd11 but not on freebsd10 (I couldn't locate build log for the failed freebsd11 system. I'll take a look if you can point me to it). I appreciate your help.

Please schedule a full CI.

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/17638/

addaleax commented:
Sorry, sure.

CI with `--use-largepages`: https://ci.nodejs.org/job/node-test-commit/22044/

suresh-srinivas commented:
We were suggested to raise Exception in `configure.py` when the option is not supported on an environment. It now shows up as tests failed. Several of the failures are due to this.. Here is an example from `https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd11-x64/20981/console`

@addaleax @uttampawar @gireeshpunathil @lundibundi what should we be doing for this?


```

python ./configure --verbose --use-largepages
23:17:34 Traceback (most recent call last):
23:17:34   File "./configure", line 28, in <module>
23:17:34     import configure
23:17:34   File "/usr/home/iojs/build/workspace/node-test-commit-freebsd/nodes/freebsd11-x64/configure.py", line 1558, in <module>
23:17:34     configure_node(output)
23:17:34   File "/usr/home/iojs/build/workspace/node-test-commit-freebsd/nodes/freebsd11-x64/configure.py", line 1008, in configure_node
23:17:34     'Large pages are supported only on Linux Systems.')
23:17:34 Exception: Large pages are supported only on Linux Systems.
```

`https://ci.nodejs.org/job/node-test-commit-linux/nodes=ubuntu1710-x64/22111/console` is failing and not really large pages related.
```

23:57:52 not ok 2347 sequential/test-fs-watch-system-limit
23:57:52   ---
23:57:52   duration_ms: 74.722
23:57:52   severity: fail
23:57:52   exitcode: 1
23:57:52   stack: |-
23:57:52     internal/child_process.js:359
23:57:52         throw errnoException(err, 'spawn');
23:57:52         ^
23:57:52     
23:57:52     Error: spawn ENOMEM
23:57:52         at ChildProcess.spawn (internal/child_process.js:359:11)
```

suresh-srinivas commented:
On Ubuntu 14.04 which doesnt have transparent pages enabled, the `printf` that we doing to `stderr` is messing with the output `https://ci.nodejs.org/job/node-test-commit-linux/nodes=ubuntu1404-64/22111/consoleFull`

We can check the Kernel version in `configure.py` through `os.uname()` and raise exception for older Linuxes. Or we could delete the printf.

```
not ok 167 parallel/test-child-process-promisified
23:28:28   ---
23:28:28   duration_ms: 1.57
23:28:28   severity: fail
23:28:28   exitcode: 1
23:28:28   stack: |-
23:28:28     Could not open file: /sys/kernel/mm/transparent_hugepage/enabled
23:28:28     /home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1404-64/test/common/index.js:662
23:28:28     const crashOnUnhandledRejection = (err) => { throw err; };
23:28:28                                                  ^
23:28:28     
23:28:28     AssertionError [ERR_ASSERTION]: Expected values to be strictly deep-equal:
23:28:28     + actual - expected
23:28:28     
23:28:28       {
23:28:28     +   stderr: 'Could not open file: /sys/kernel/mm/transparent_hugepage/enabled\n',
23:28:28     -   stderr: '',
23:28:28         stdout: '42\n'
23:28:28       }
23:28:28         at execFile.then.common.mustCall (/home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1404-64/test/parallel/test-child-process-promisified.js:18:12)
23:28:28         at /home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1404-64/test/common/index.js:349:15
23:28:28         at process._tickCallback (internal/process/next_tick.js:68:7)
```

lundibundi commented:
@suresh-srinivas That's completely ok if build fails on unsupported platforms, don't worry about this.

Though I'm still concerned with strange `configure` behavior on freebsd and osx (mentioned in https://github.com/nodejs/node/pull/22079#issuecomment-426386196).

Here are the logs
* OSX10 - https://ci.nodejs.org/job/node-test-commit-osx/nodes=osx1010/21671/console
* OSX11 - https://ci.nodejs.org/job/node-test-commit-osx/21687/nodes=osx1011/console
* freebsd10-64 - https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd10-64/20965/console
* freebsd11-x64 - https://ci.nodejs.org/job/node-test-commit-freebsd/20981/nodes=freebsd11-x64/console

/cc @uttampawar

suresh-srinivas commented:
> @suresh-srinivas That's completely ok if build fails on unsupported platforms, don't worry about this. We may later start a clean build only on supported ones, it's just easier to use run-all sort of runner.

Thanks. That will be helpful. Let us know what we need to look into.

> 
> Though I'm still concerned with strange `configure` behavior on freebsd and osx (mentioned in [#22079 (comment)](https://github.com/nodejs/node/pull/22079#issuecomment-426386196)).

This is no longer an issue since we modified `configure.py` with tighter checks. See the newest logs from `https://ci.nodejs.org/job/node-test-commit-osx/nodes=osx1011/21687/console`, this gets a configure error and will not build. Similarly `https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd11-x64/20981/console`


> 
> Here are the logs
> 
> * OSX10 - https://ci.nodejs.org/job/node-test-commit-osx/nodes=osx1010/21671/console
> * OSX11 - https://ci.nodejs.org/job/node-test-commit-osx/21687/nodes=osx1011/console
> * freebsd10-64 - https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd10-64/20965/console
> * freebsd11-x64 - https://ci.nodejs.org/job/node-test-commit-freebsd/20981/nodes=freebsd11-x64/console
> 
> /cc @uttampawar

lundibundi commented:
> This is no longer an issue since we modified configure.py with tighter checks. See the newest logs from https://ci.nodejs.org/job/node-test-commit-osx/nodes=osx1011/21687/console, this gets a configure error and will not build. Similarly https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd11-x64/20981/console

I looked at the newest CI 
> CI with --use-largepages: ci.nodejs.org/job/node-test-commit/22044

That's exactly my question, why does it successfully builds and runs on freebsd10-64 and osx10 but fails to be built with 'Supported only on linux' on freebsd11-x64 and osx11, it that okay (why was freebsd10 counted as Linux but freebsd 11 is not)?

uttampawar commented:
> @suresh-srinivas
> 
> > This is no longer an issue since we modified configure.py with tighter checks. See the newest logs from https://ci.nodejs.org/job/node-test-commit-osx/nodes=osx1011/21687/console, this gets a configure error and will not build. Similarly https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd11-x64/20981/console
> 
> I looked at the newest CI
> 
> > CI with --use-largepages: ci.nodejs.org/job/node-test-commit/22044
> 
> That's exactly my question, why does it successfully builds and runs on freebsd10-64 and osx10 but fails to be built with 'Supported only on linux' on freebsd11-x64 and osx11, it that okay (why was freebsd10 counted as Linux but freebsd 11 is not)?

@lundibundi 
It looks like build# 29065  (on freebsd10 - https://ci.nodejs.org/job/node-test-commit-freebsd/nodes=freebsd10-64/20965/) and build# 21671 (on osx1010- https://ci.nodejs.org/job/node-test-commit-osx/nodes=osx1010/21671/) picked up an earlier commit so it succeeded,

lundibundi commented:
CI with `--use-largepages`: https://ci.nodejs.org/job/node-test-commit/22063/

addaleax commented:
New CI: https://ci.nodejs.org/job/node-test-commit/22154/

suresh-srinivas commented:
> New CI: https://ci.nodejs.org/job/node-test-commit/22154/

Thank you Anna. All the failures on platforms other than Linux-64 are due to `configure ` error for not supported on the platform.

I looked through all the rest of the Linux, the failures were on these. Looks like the alpine-last-latest-x64 error is a `SEGV` and the centos6-64-gcc6 is a compile error (with `MADV_HUGEPAGE`)
```

https://ci.nodejs.org/job/node-test-commit-linux/22222/nodes=alpine-last-latest-x64/consoleFull

not ok 241 parallel/test-child-process-spawnsync-kill-signal
11:19:37   ---
11:19:37   duration_ms: 1.700
11:19:37   severity: fail
11:19:37   exitcode: 1
11:19:37   stack: |-
11:19:37     assert.js:86
11:19:37       throw new AssertionError(obj);
11:19:37       ^
11:19:37     
11:19:37     AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
11:19:37     + actual - expected
11:19:37     
11:19:37     + 'SIGSEGV'
11:19:37     - 'SIGTERM'
11:19:37           ^
11:19:37         at Object.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/alpine-last-latest-x64/test/parallel/test-child-process-spawnsync-kill-signal.js:42:12)
11:19:37         at Module._compile (internal/modules/cjs/loader.js:706:30)
11:19:37         at Object.Module._extensions..js (internal/modules/cjs/loader.js:717:10)
11:19:37         at Module.load (internal/modules/cjs/loader.js:604:32)
11:19:37         at tryModuleLoad (internal/modules/cjs/loader.js:543:12)
11:19:37         at Function.Module._load (internal/modules/cjs/loader.js:535:3)
11:19:37         at Function.Module.runMain (internal/modules/cjs/loader.js:759:12)
11:19:37         at startup (internal/bootstrap/node.js:303:19)
11:19:37         at bootstrapNodeJSCore (internal/bootstrap/node.js:866:3)
11:19:37   ...

https://ci.nodejs.org/job/node-test-commit-linux/22222/nodes=centos6-64-gcc6/console


../src/node_large_page.cc: In function ‘int node::MoveTextRegionToLargePages(const node::text_region&)’:
11:24:30 ../src/node_large_page.cc:219:29: error: ‘MADV_HUGEPAGE’ was not declared in this scope
11:24:30    ret = madvise(tmem, size, MADV_HUGEPAGE);

https://ci.nodejs.org/job/node-test-commit-linux/22222/nodes=debian8-64/console
not ok 2338 sequential/test-fs-watch-system-limit
11:28:26   ---
11:28:26   duration_ms: 5.545
11:28:26   severity: fail
11:28:26   exitcode: 1
11:28:26   stack: |-
11:28:26     internal/child_process.js:358
11:28:26         throw errnoException(err, 'spawn');
11:28:26         ^
11:28:26     
11:28:26     Error: spawn ENOMEM
11:28:26         at ChildProcess.spawn (internal/child_process.js:358:11)
11:28:26         at Object.spawn (child_process.js:538:9)
11:28:26         at spawnProcesses (/home/iojs/build/workspace/node-test-commit-linux/nodes/debian8-64/test/sequential/test-fs-watch-system-limit.js:20:32)
11:28:26         at Timeout.setTimeout [as _onTimeout] (/home/iojs/build/workspace/node-test-commit-linux/nodes/debian8-64/test/sequential/test-fs-watch-system-limit.js:33:7)
11:28:26         at ontimeout (timers.js:464:11)
11:28:26         at tryOnTimeout (timers.js:336:5)
11:28:26         at listOnTimeout (timers.js:310:5)
11:28:26         at processTimers (timers.js:267:5)
11:28:26   ...

https://ci.nodejs.org/job/node-test-commit-linux/22222/nodes=fedora-last-latest-x64/consoleFull

FATAL: command execution failed
11:24:49 java.nio.channels.ClosedChannelException
11:24:49 	at org.jenkinsci.remoting.protocol.NetworkLayer.onRecvClosed(NetworkLayer.java:154)
11:24:49 	at org.jenkinsci.remoting.protocol.impl.NIONetworkLayer.ready(NIONetworkLayer.java:142)

https://ci.nodejs.org/job/node-test-commit-linux/22222/nodes=fedora-latest-x64/console

FATAL: command execution failed
11:25:08 java.nio.channels.ClosedChannelException

https://ci.nodejs.org/job/node-test-commit-linux/22222/nodes=ubuntu1404-64/consoleFull

FATAL: command execution failed
11:29:09 java.nio.channels.ClosedChannelException

https://ci.nodejs.org/job/node-test-commit-linux/22222/nodes=ubuntu1604-64/console

FATAL: command execution failed
11:30:21 java.nio.channels.ClosedChannelException



```

lundibundi commented:
Full CI: https://ci.nodejs.org/job/node-test-commit/22190/

Wanted to make sure it will properly fail on both freebsd 10/11 and macos (as discussed in prev messages).

Also, will rerun those that failed with `ClosedChannelException`.

mhdawson commented:
On request from @gabrielschulhof uname for the machine on which we see a failure

```
[root@test-softlayer-centos6-x64-1 ~]# uname -a
Linux test-softlayer-centos6-x64-1 2.6.32-696.28.1.el6.x86_64 #1 SMP Wed May 9 23:09:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

gabrielschulhof commented:
New CI with `--use-largepages`: https://ci.nodejs.org/job/node-test-commit-linux/22263/

bmeurer commented:
Coming to the discussion late, so not sure this has been discussed: But how about putting this into `ld.so` directly? It seems strange that the application redoes some part of the linker's job here. It seems like this way we loose the ability to resolve symbols in the node binary, as all the addresses are different from the ones found by looking into the DWARF info? Or am I missing something here?

suresh-srinivas commented:
> Coming to the discussion late, so not sure this has been discussed: But how about putting this into `ld.so` directly? It seems strange that the application redoes some part of the linker's job here. It seems like this way we loose the ability to resolve symbols in the node binary, as all the addresses are different from the ones found by looking into the DWARF info? Or am I missing something here?

The addresses are the same, it is only remapped to 2M pages. There is a linker script that is used in the solution. Could you be specific on what you are not able to do? I am able to use `gdb` to run the new binary and put breakpoints etc.

refack commented on ld.implicit.script:
```diff
@@ -0,0 +1,8 @@
+  SECTIONS {
```
Could you move this to `/src/`?

uttampawar replied:
@refack Any particular reason do you feel to create new directory? It is new license or something else?

refack replied:
In general I'm in favor of more fine grained structuring for the tree. This seems like a cohesive module, so it seems like a good candidate.
And yes, that it is under a different license makes it unique. Might even need to be a `/dep/`.
Also having a readme in that directory will help document this for posterity.

lundibundi replied:
+1 to change, haven't noticed it was in the root of the repo.
I'd say at least to simply group the files in one place, this will make maintenance and understanding of the implementation easier IMO.

addaleax replied:
> Might even need to be a `/dep/`.

Please don’t move files to `deps/` that are not imported and updated from an upstream repo. 

> And yes, that it is under a different license makes it unique.

It is not under a different license, and therefore not unique in that respect.

refack replied:
> It is not under a different license, and therefore not unique in that respect.

Yes it is, see https://github.com/nodejs/node/pull/22079#discussion_r207097339
Neither of us are layers, but I'd rather be safe than sorry.

refack replied:
As simple solution might be to put it in `/src/large_pages/` and exclude that directory in our LICENSE.

suresh-srinivas replied:
Moving it to a different directory and containing it sounds good to me. @uttampawar and I will work on it.

refack replied:
Thank you 🎉

uttampawar replied:
@refack Addressed in the latest updated PR. Can you explain your thoughts about LICENSE file content, please?

refack commented on node.gypi:
```diff
@@ -194,7 +194,18 @@
+              and target_arch=="x64" \
+              and node_use_large_pages=="true"', {
+              'ldflags': [
+                 '-Wl,-T <(PRODUCT_DIR)/../../ld.implicit.script',
```
This should be:
```
                '-Wl,-T src/ld.implicit.script',
```
Using `src/` after you move `ld.implicit.script` to `src`

uttampawar replied:
@refack wouldn't this be <(PRODUCT_DIR)/../../src/large_pages/ld.implicit.script?

refack commented on node.gypi:
```diff
@@ -194,7 +194,18 @@
           },
           'conditions': [
-            ['OS!="aix" and node_shared=="false"', {
+            ['OS=="linux" and node_shared=="false" \
```
No need for `/`

uttampawar replied:
@refack Addressed in the latest updated PR.

refack commented on src/node.cc:
```diff
@@ -68,6 +68,7 @@
 #ifdef NODE_ENABLE_VTUNE_PROFILING
 #include "../deps/v8/src/third_party/vtune/v8-vtune.h"
 #endif
+#include "node_large_page.h"
```
This should also be conditional on `NODE_ENABLE_LARGE_CODE_PAGES`

uttampawar replied:
@refack Addressed in the latest updated PR. Can you elaborate more on your suggestion about readme and license file?

refack commented:
/CC @nodejs/build-files @nodejs/python

refack commented on node.gyp:
```diff
@@ -578,6 +578,19 @@
+          'conditions': [
+            [ 'OS=="linux"', {
+              'sources': [
+                'src/node_large_page.cc'
```
could you add 'src/large_pages/node_large_page.h' as well

uttampawar replied:
@refack Addressed in the latest updated PR.

refack commented on node.gyp:
```diff
@@ -578,6 +578,19 @@
+          # Other implementations are possible but not currently supported.
+          #
+          'conditions': [
+            [ 'OS=="linux"', {
```
Could you move this condition as an `and` to L581

uttampawar replied:
@refack Addressed in the latest updated PR.

bmeurer commented:
> The addresses are the same, it is only remapped to 2M pages. There is a linker script that is used in the solution. Could you be specific on what you are not able to do? I am able to use `gdb` to run the new binary and put breakpoints etc.

Ah, I missed that part. Thanks for the explanation.

Can you elaborate on why you don't want to put this into `ld.so`? It seems like we're cleaning up after the linker here, so wouldn't it be more effective to fix the linker directly?

suresh-srinivas commented:
> > The addresses are the same, it is only remapped to 2M pages. There is a linker script that is used in the solution. Could you be specific on what you are not able to do? I am able to use `gdb` to run the new binary and put breakpoints etc.
> 
> Ah, I missed that part. Thanks for the explanation.
> 
> Can you elaborate on why you don't want to put this into `ld.so`? It seems like we're cleaning up after the linker here, so wouldn't it be more effective to fix the linker directly?

`ld.so` is for linking. We are using `madvise` using transparent huge pages at runtime to get them into 2M pages. Take a look at [Transparent Huge Pages](https://www.kernel.org/doc/Documentation/vm/transhuge.txt) documentation. I dont believe there is a way to do this at compile or link time. Another implementation I can think of is using the [HugeTLBFS](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt) and even that needs `mmap` even if you allocate it.

addaleax commented:
@suresh-srinivas I think @bmeurer’s question makes sense – for non-static binaries, `ld.so` is the program that maps the executable and its sections into memory, not just for linking symbols. It seems like a better place to implement this than doing it for individual programs like Node.js, on first thought.

suresh-srinivas commented:
> @suresh-srinivas I think @bmeurer’s question makes sense – for non-static binaries, `ld.so` is the program that maps the executable and its sections into memory, not just for linking symbols. It seems like a better place to implement this than doing it for individual programs like Node.js, on first thought.

Yes if we could somehow specify or automatically to load a group of addresses into 2M pages that would be ideal.  However the Linux operating systems only provides a more explicit interface for huge pages to user space. It is up to application developers and system administrators to use it.

 I had originally proposed using the [`libhugetlbfs`](https://github.com/libhugetlbfs/libhugetlbfs) library which had abstracted the complexity and just needed  linking with that library, but got community feedback to not use that approach, which is why I developed all this code.

addaleax commented:
At the risk of sounding a bit repetitive… is there a reason why `ld.so` can’t do this? From the kernel perspective, it’s part of the application code, and since it is the part that originally loads the `node` binary’s sections into memory (or the ones of any other dynamically linked program), it seems like a well-suited place to do this?

suresh-srinivas commented:
> At the risk of sounding a bit repetitive… is there a reason why `ld.so` can’t do this? From the kernel perspective, it’s part of the application code, and since it is the part that originally loads the `node` binary’s sections into memory (or the ones of any other dynamically linked program), it seems like a well-suited place to do this?

Let me take an example. If we take the case of a simple statically linked elf file, the application has a fixed load address (assigned by the linker) that the kernel knows. The load process consists of making the application available at the correct address space of the newly created process and transferring control to it.  The default page size is 4K and that is what the kernel will use in the newly created process. There isnt anything that `ld.so`  is doing here.  If the default size of the kernel happened to be 2M, then it will use that size.

Hope it is clear. What we are trying to change the default of the kernel for the `.text` segment to be 2M pages and have to use the mmap and madvise to do this.

addaleax commented:
@suresh-srinivas Yes, if the answer is “because it doesn’t work for statically linked executables and that’s important to us” then it’s perfectly clear why you are doing it here. That just wasn’t stated anywhere. :)

suresh-srinivas commented:
> @suresh-srinivas Yes, if the answer is “because it doesn’t work for statically linked executables and that’s important to us” then it’s perfectly clear why you are doing it here. That just wasn’t stated anywhere. :)

Yes, thanks for restating it. It was originally discussed in the [Large Page Issue](https://github.com/nodejs/node/issues/16198) that  the statically linked node itself is the predominant code in the booted up node. We can add it to the documentation to make sure the assumptions are clear.

addaleax commented:
> It was originally discussed in the Large Page Issue that the statically linked node itself is the predominant code in the booted up node.

@suresh-srinivas Sorry, but that makes things more confusing again :smile: I can’t find that in the linked issue. The closest thing is that y’all seemed to agree that the main focus should be on the sections coming from the `node` binary itself, but that has nothing to do with whether the binary is statically linked or not, right? Calling static linking “predominant” is surprising – I’d personally estimate that > 99.99 % of Node.js executions come from dynamically linked executables.

addaleax commented:
@suresh-srinivas I feel like we’re talking past each other… so, trying to clear things up:

- I don’t think there’s disagreement that the `.text` segment of `node` is something that’s reasonable to focus on when working on this issue in Node.js.
- What is not clear to me is whether statically vs dynamically linked executables make a difference.
  - If yes, why? Statically linked `node` binaries are supported, but rare, and need explicit opt-in during `./configure`.
  - If not, is there a reason not to also try to approach this issue in `ld.so` in a more generic way, so that all dynamically linked executables & shared libraries on a system can benefit from it?

suresh-srinivas commented:
> @suresh-srinivas I feel like we’re talking past each other… so, trying to clear things up:

Yes, thanks for being patient. I feel like I am repeating myself.

> 
> * I don’t think there’s disagreement that the `.text` segment of `node` is something that’s reasonable to focus on when working on this issue in Node.js.

Ok good that we agree on this.

> * What is not clear to me is whether statically vs dynamically linked executables make a difference.
>   
>   * If yes, why? Statically linked `node` binaries are supported, but rare, and need explicit opt-in during `./configure`.

I feel that best performanc will come from this, where all the libraries are linked in. In the above case both the node binary and the libpthread will benefit from 2M pages, but the Dynamically generated code will not benefit from 2M pages.

>   * If not, is there a reason not to also try to approach this issue in `ld.so` in a more generic way, so that all dynamically linked executables & shared libraries on a system can benefit from it?

I am not sure how this would work and if there would be a way to specify which part to use 2M pages for. I can check with some people who are knowledgable about ld. I also dont know how we would use this.

I also think that the JITted code will not be able to use the 2M pages. That was one of the other next steps that I wanted to look into for v8 or chakra.

Another approach to benefit more programs is to make this functionality into a library and during the library init to do the remapping. It would be an opt in and linking with the library.

refack commented on node.gypi:
```diff
@@ -312,6 +301,9 @@
                    '-Wl,-z,now' ]
     }],
+    [ 'OS=="linux" and target_arch=="x64" and node_use_large_pages=="true"', {
+      'ldflags': [ '-Wl,-T <(PRODUCT_DIR)/../../src/large_pages/ld.implicit.script' ]
```
Simple `'-Wl,-T src/large_pages/ld.implicit.script'` doesn't work?
Or `'-Wl,-T', 'src/large_pages/ld.implicit.script'` doesn't work?

suresh-srinivas replied:
Yes we tried `-Wl,-T src/large_pages/ld.implicit.script` and that didnt work. Looks like `-T` needs full path or some `-L` with directories to search for.

refack replied:
Thank you.
It's not going to be a problem with `make` and linux, but FTR it will not work with `ninja`, `msvs` or cross compilers.

suresh-srinivas commented:
> @suresh-srinivas
> 
> > I am not sure how this would work
> 
> `ld.so` is the script that, at least under all Linuxes that I know about, runs as the very first thing in a process, even before the entry point of the main executable (e.g. `node`) – if it is dynamically linked, i.e. not compiled with `--static`. At that point, only itself and the main executable will have been loaded into memory by the kernel. For the main executable, it could re-map the pages in a similar way to what we do here – without the need to worry about inlining effects or anything, since the code in the target pages isn’t even running yet – and for shared libraries, it could specify large pages when `mmap()`ing them (since loading & linking them is the other part of its job).

Ok got it. Would we be able to specify the stub address in this scheme? Where is `ld.so` loaded and would we be sure that it is outside of the executable address? I am using a trick in `ld.implicit.script` to specify an address range outside of the executable for the code that does the move as it itself should not move.

> 
> > and if there would be a way to specify which part to use 2M pages for.
> 
> That’s a more interesting question. I guess it depends on whether it’s possible to tell for a given executable in advance whether it makes sense to load some of its sections as large pages – I would assume that it makes sense whenever a section is a few megabytes in size, but I could totally be wrong. (In any case, this is your domain of expertise. )

I experimentally figured this out for `node` applications. I found that majority of the TLB misses are coming from the `.text` segment of node itself.

> 
> I’m also wondering if you could explain why you think this makes less sense for JIT code – I don’t think it has anything to do with the discussion here, but I imagine the reason would provide some interesting insights. :)

The JITted code is on the heap, dynamically allocated and then marked as execute after the code is generated. `ld.so` is not involved in this, only the JIT is. The JIT will need to allocate or copy the generated code to 2M pages. They would need to use similar code to here. 

Also we have not talked about putting JavaScript data in large pages. This would be needed if an application has a lot of DTLB misses. For example Java JVM has `-XX:+UseLargePages` to allocate heap in 2M pages.

> 
> > I can check with some people who are knowledgable about ld.
> 
> Just so we don’t misunderstand, this is mostly about `ld.so` (the dynamic linker/executable interpreter provided by the system libc), not `ld` (the linker as part of the compiler toolchain).
> 
> (So, yes, this would require patching libc source code. That’s also a totally valid reason not to do this; on the other hand, if this gives even a slight performance boost, it might be worth a ton of work, because of how wide-reaching the improvements could be.)

Ok got it. The approach we are advocating works for applications that have high ITLB misses. I dont know if this would be something more general.

> 
> > I also dont know how we would use this.
> 
> I guess that depends on all of the above. :)

We also will need to remember that any code outside the system will need very clear dependency and older systems that have the right Linux kernel but not the updated `ld.so` would not benefit. Given that we recently made some changes to libc to handle AVX3 Licensing, it has taken more than a year and very few backports.

suresh-srinivas commented:
It looks like the linter is failing. Could someone help diagnose? I cant see the exact error? locally make linter works. From high level looks like something wrong with title-length and subsystem.
@gabrielschulhof @addaleax @lundibundi 

https://travis-ci.com/nodejs/node/jobs/151975184

```
9ed4646df05b9593075bb0cd0b3ab420bae482a5...ef78b73baf6eaf5350adfe7dc1131249651efedf
npx: installed 13 in 1.748s
  ✖  2065fea8a4f3e4b79bf1b1e6a675f24349dbbd12
     ✔  0:0      skipping fixes-url                        fixes-url
     ✔  0:0      blank line after title                    line-after-title
     ✔  0:0      line-lengths are valid                    line-length
     ✖  0:0      Missing subsystem.                        subsystem
     ✔  0:0      Title is formatted correctly.             title-format
     ✖  0:72     Title must be <= 72 columns.              title-length

```

mmarchini commented:
Commits are not following [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines).

refack commented on node.gypi:
```diff
@@ -301,6 +301,9 @@
                    '-Wl,-z,now' ]
     }],
+    [ 'OS=="linux" and target_arch=="x64" and node_use_large_pages=="true"', {
+      'ldflags': [ '-Wl,-T <(PRODUCT_DIR)/../../src/large_pages/ld.implicit.script' ]
```
```suggestion
      'ldflags': [
        '-Wl,-T',
        '<!(realpath src/large_pages/ld.implicit.script)',
      ]
```
This should work, and be cross platform

uttampawar replied:
@refack Thanks for the suggestion. I'll make the change and update the PR.

uttampawar commented:
I've updated out PR with another attempt to fix the linter error discovered with previous build. It had two failures but second one doesn't look like due to large pages related.

uttampawar commented:
Travis CI failed again with following message(s),
$ git log $TRAVIS_COMMIT_RANGE --pretty=format:'%h' --no-merges | tail -1 | xargs npx core-validate-commit --no-validate-metadata
npx: installed 13 in 3.257s
  ✖  9d7895c567e8f38abfff35da1b6d6d6a0a06f9aa
     ✔  0:0      skipping fixes-url                        fixes-url
     ✔  0:0      blank line after title                    line-after-title
     ✔  0:0      line-lengths are valid                    line-length
     ✖  0:0      Missing subsystem.                        subsystem
     ✔  0:0      Title is formatted correctly.             title-format
     ✔  0:0      Title is <= 50 columns.                   title-length

$ git log --pretty=oneline shows message with "src" as subsystem as shown below,
60dcd05f2c623db171d181345d6735aaccaa1833 src: Initial Large Page (2M) support for node.js

I missing something for sure but don't really know what? I would appreciate if anyone can help me in resolving this error.

mmarchini commented:
Looks like an issue with the linter this time, unrelated to your commit.

CI: https://ci.nodejs.org/job/node-test-pull-request/17973/
Linux CI with `--use-largepages`: https://ci.nodejs.org/job/node-test-commit-linux/22486/

refack commented:
Landed as bf7ed80475cfa7ae1ab6086c325f57557b75dffe 🎉 
Thank you for the contribution, and the continued engagement!

uttampawar commented:
@lundibundi @refack @addaleax @gireeshpunathil Thank you all for your feedback, support and importantly patience while we work through various issues for over approx. 3 months.

suresh-srinivas commented:
Yes thanks @uttampawar @gireeshpunathil @addaleax @refack @lundibundi @mmarchini @gabrielschulhof @mhdawson  @joyeecheung for all of your feedback and support during this process. Right from when the  [Using 2M pages for code](https://github.com/nodejs/node/issues/16198) issue was opened, through all the technical discussions, starting the CI's and many more. Appreciate it. We learned a tremendous amount during this process and couldnt have done it without all of you.

refack commented:
> Linux CI with `--use-largepages`: [ci.nodejs.org/job/node-test-commit-linux/22486](https://ci.nodejs.org/job/node-test-commit-linux/22486/)

Now we (the build WG) need to formulate a way to test this in CI. ATM we don't have a way to select linux platforms by kernel version...

addaleax commented:
Should this be backported to `v10.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.

suresh-srinivas commented:
> Should this be backported to `v10.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.

Thanks, yes we will work on raising a PR @uttampawar
